sos/report/plugins/ceph_osd.py

```diff
@@ -39,9 +40,47 @@ def setup(self):
 
         self.add_cmd_output([
             "ceph-disk list",
-            "ceph-volume lvm list",
         ])
 
+        ceph_cmds = [
+           "bluestore bluefs available",
+           "config diff",
+           "config show",
+           "dump_blacklist",
+           "dump_blocked_ops",
+           "dump_historic_ops_by_duration",
+           "dump_historic_slow_ops",
+           "dump_mempools",
+           "dump_ops_in_flight",
+           "dump_op_pq_state",
+           "dump_osd_network",
+           "dump_reservations",
+           "dump_watchers",
+           "log dump",
+           "perf dump",
+           "objecter_requests",
+           "ops",
+           "status"
+        ]
+
+        osd_ids = []
+        output = self.collect_cmd_output('ceph-volume lvm list --format json')
```

Comment:
There have been a number of performance issues with ceph-volume in the past. I do not recommend using this command to build a list of osd ids. 

It is a bit uglier, but you can build a list of ids by parsing `mount` output. As an example, this is the command systemd uses to generate a list of the ceph-osd services that need to be started:
`sudo mount|grep /var/lib/ceph/osd/ceph-|awk "{print \$3}"|cut -d- -f2|sort -n|xargs`

Reply:
An equally ugly but slightly more efficient version of it can be
```
sudo mount|awk -v p=/var/lib/ceph/osd/ceph- '$0~p{split($3, x, "-"); printf x[2]" "}'
```
(sorting isn't necessary here).

Reply:
Hi @pponnuvel  , @hillpd 

thank you for the review and suggestions!

The collect_cmd_output() method provided by sosreport does not allow pipes. So we need to get the output of a command, then implement the output extraction in the sos plugin.

Another approach might be to extract it from output of ps -u ceph -o args, and extract the id numbers that way (The ceph-osd process belong to the user ceph). 

output of "ps -u ceph -o args" looks like - 

COMMAND
/usr/bin/ceph-osd -f --cluster ceph --id 0 --setuser ceph --setgroup ceph
/usr/bin/ceph-osd -f --cluster ceph --id 1 --setuser ceph --setgroup ceph
/usr/bin/ceph-osd -f --cluster ceph --id 2 --setuser ceph --setgroup ceph

If we use this approach, we would not collect data for an OSD that is stopped by admin, or even down/crashed, possibly saving the plugin some time. Please let me know what you think.

Reply:
Chiming in: pipes and bash foo in general aren't allowed in plugins.

You could do something like the following, if `ceph-volume lvm list` is not desired:

```python
out = self.collect_cmd_output('mount')
mounts = [
    m.split()[2] for m in out['output'].splitlines() if 'ceph/osd/ceph-' in m
]
```

However, the actual collection of `ceph-volume lvm list` seems important as it's been in sos for a while. What kind of performance concerns are there? Is it just related to the `ceph-volume` command, or does this affect the entire host? If we're going to keep collecting the output of `ceph-volume lvm list`, I have no issues with using it to build the list.

We have several safe guards in place to prevent any under-performing command collection from impacting the rest of the plugin, or the report as a whole, too much. At worst, we'd hit a timeout and report that timeout to the user, then continue on as normal.

Reply:
@nkshirsagar 
I think parsing either `ps` or `mount` output is a reasonable approach.

@TurboTurtle
I'm not suggesting that we stop collecting `ceph-volume lvm list`. We should just avoid using this command to build the `osd_ids` list. A time-out would prevent all of the `ceph_cmds` from being collected on lines 78 and 82.

> What kind of performance concerns are there? Is it just related to the `ceph-volume` command, or does this affect the entire host?

My concern is that the `ceph-volume` command may be prone to time-out on hosts running older Ceph code levels that sosreport can still target. There are several bugs I'm concerned about: [lp#1850754](https://bugs.launchpad.net/ubuntu/+source/ceph/+bug/1850754) and [lp#1908375](https://bugs.launchpad.net/ubuntu/+source/ceph/+bug/1908375).